contracts/src/v0.8/dev/automation/2_1/AutomationRegistrar2_1.sol

```diff
@@ -0,0 +1,518 @@
+   */
+  function setRegistrationConfig(address keeperRegistry, uint96 minLINKJuels) public onlyOwner {
+    s_config = RegistrarConfig({minLINKJuels: minLINKJuels, keeperRegistry: IKeeperRegistryMaster(keeperRegistry)});
+    emit ConfigChanged(keeperRegistry, minLINKJuels);
+  }
+
+  /**
+   * @notice owner calls to set the cofig for this upkeep type
+   * @param upkeepTypeConfig config to set
+   */
+  function setUpkeepTypeConfig(UpkeepTypeConfig memory upkeepTypeConfig) public onlyOwner {
+    s_upkeepTypeConfigs[upkeepTypeConfig.upkeepType].autoApproveType = upkeepTypeConfig.autoApproveType;
+    s_upkeepTypeConfigs[upkeepTypeConfig.upkeepType].autoApproveMaxAllowed = upkeepTypeConfig.autoApproveMaxAllowed;
+  }
+
+  /**
+   * @notice owner calls this function to set allowlist status for senderAddress
+   * @param senderAddress senderAddress to set the allowlist status for
+   * @param allowed true if senderAddress needs to be added to allowlist, false if needs to be removed
+   */
+  function setAutoApproveAllowedSender(address senderAddress, bool allowed) external onlyOwner {
+    s_autoApproveAllowedSenders[senderAddress] = allowed;
+
+    emit AutoApproveAllowedSenderSet(senderAddress, allowed);
+  }
+
+  /**
+   * @notice read the allowlist status of senderAddress
+   * @param senderAddress address to read the allowlist status for
+   */
+  function getAutoApproveAllowedSender(address senderAddress) external view returns (bool) {
+    return s_autoApproveAllowedSenders[senderAddress];
+  }
+
+  /**
+   * @notice read the current registration configuration
+   */
+  function getRegistrationConfig() external view returns (address keeperRegistry, uint256 minLINKJuels) {
+    RegistrarConfig memory config = s_config;
+    return (address(config.keeperRegistry), config.minLINKJuels);
+  }
+
+  /**
+   * @notice read the config for this upkeep type
+   * @param upkeepType upkeep type to read config for
+   */
+  function getUpkeepTypeConfig(uint8 upkeepType) external view returns (UpkeepRegistrationState memory) {
+    return s_upkeepTypeConfigs[upkeepType];
+  }
+
+  /**
+   * @notice gets the admin address and the current balance of a registration request
+   */
+  function getPendingRequest(bytes32 hash) external view returns (address, uint96) {
+    PendingRequest memory request = s_pendingRequests[hash];
+    return (request.admin, request.balance);
+  }
+
+  /**
+   * @notice Called when LINK is sent to the contract via `transferAndCall`
+   * @param sender Address of the sender transfering LINK
+   * @param amount Amount of LINK sent (specified in Juels)
+   * @param data Payload of the transaction
+   */
+  function onTokenTransfer(
+    address sender,
+    uint256 amount,
+    bytes calldata data
+  )
+    external
+    override
+    onlyLINK
+    permittedFunctionsForLINK(data)
+    isActualAmount(amount, data)
+    isActualSender(sender, data)
+  {
+    if (data.length < 292) revert InvalidDataLength();
```

Comment:
i see this is the same as old registrar. does this need to increase as new fields have been added within data

Reply:
I'm not entirely sure why we have this check in the first place... I noticed we do it on other `transferAndCall` functions but not all. Not sure if the length check is to address a security concern (if so, idk what) or to provide better error messages.

But! I'm not sure it really matters because I think it's redundant. The `isActualAmount` and `isActualSender` modifiers decode the data into the expected types of `register()`(which requires a minimum length). I'm pretty sure there's no test we can write that would trigger the `InvalidDataLength()` error. So I just removed the length check all together.

Reply:
iirc we had different code in `isActualAmount` and `isActualSender` before when we were firstly fetching value from an offset instead of using abi decode. Now makes sense, because we are using abi decode it automatically would also do a length check there